provisioning.go

```diff
@@ -0,0 +1,370 @@
+//
+// This function will create a pair of NV indices used for locking access to sealed key objects, if necessary. These indices will be
+// created at 0x01801100 and 0x01801101. This requires knowledge of the authorization value for the storage hierarchy. If called with
+// mode set to ProvisionModeClear, or if called just after clearing the TPM via the physical presence interface, the authorization
+// value for the storage hierarchy will be empty at the point that it is required. If called with any other mode and if the
+// authorization value for the storage hierarchy has previously been set, it will need to be provided by calling
+// TPMConnection.OwnerHandleContext().SetAuthValue() prior to calling this function. If the wrong value is provided for the storage
+// hierarchy authorization, then a AuthFailError error will be returned. If the correct authorization value is not known and new NV
+// indices need to be created, then the only way to recover from this is to call the function with mode set to ProvisionModeClear. If
+// there are already NV indices defined at either of these handles but they don't meet the requirements of this function, a
+// TPMResourceExistsError error will be returned. In this case, the caller will either need to manually undefined these using
+// TPMContext.NVUndefineSpace, or clear the TPM.
+//
+// If mode is not ProvisionModeWithoutLockout, the authorization value for the lockout hierarchy will be set to newLockoutAuth.
+func ProvisionTPM(tpm *TPMConnection, mode ProvisionMode, newLockoutAuth []byte) error {
+	status, err := ProvisionStatus(tpm)
+	if err != nil {
+		return xerrors.Errorf("cannot determine the current status: %w", err)
+	}
+
+	// Create an initial session for HMAC authorizations
+	session, err := tpm.StartAuthSession(nil, nil, tpm2.SessionTypeHMAC, nil, defaultSessionHashAlgorithm, nil)
+	if err != nil {
+		return xerrors.Errorf("cannot start session: %w", err)
+	}
+	defer tpm.FlushContext(session)
+
+	session.SetAttrs(tpm2.AttrContinueSession)
+
+	if mode == ProvisionModeClear {
+		if status&AttrOwnerClearDisabled > 0 {
+			return ErrTPMClearRequiresPPI
+		}
+
+		if err := tpm.Clear(tpm.LockoutHandleContext(), session); err != nil {
+			switch {
+			case isAuthFailError(err):
+				return AuthFailError{tpm2.HandleLockout}
+			case isLockoutError(err):
+				return ErrTPMLockout
+			}
+			return xerrors.Errorf("cannot clear the TPM: %w", err)
+		}
+
+		status = 0
+	}
+
+	// Provision an endorsement key
+	ekContext, err := tpm.CreateResourceContextFromTPM(ekHandle)
+	if err == nil {
+		if _, err := tpm.EvictControl(tpm.OwnerHandleContext(), ekContext, ekHandle, session); err != nil {
+			if isAuthFailError(err) {
+				return AuthFailError{tpm2.HandleOwner}
+			}
+			return xerrors.Errorf("cannot evict existing object at handle required by endorsement key: %w", err)
+		}
+	} else if _, notFound := err.(tpm2.ResourceUnavailableError); !notFound {
```

Comment:
(nitpick) this feels a bit unusual, most go code hanldes the `err != nil` case first which is I think would also make this easier to read. Something like:
```
        if err != nil {
		if _, notFound := err.(tpm2.ResourceUnavailableError); !notFound {
			return xerrors.Errorf("cannot create context for object at handle required by endorsement key: %w", err)
		}
	}
	if _, err := tpm.EvictControl(tpm.OwnerHandleContext(), ekContext, ekHandle, session); err != nil {
		if isAuthFailError(err) {
			return AuthFailError{tpm2.HandleOwner}
		}
		return xerrors.Errorf("cannot evict existing object at handle required by endorsement key: %w", err)
	}
```
(or even the first line with a helper: `if err != nil && !tpm2.IsNotFound(err) {` (iirc xerrors also provides something for this).

Reply:
I can reorder it, but the logic here wouldn't quite work right - TPMContext.EvictControl should only run if TPMContext.CreateResourceContextFromTPM succeeded (it will just return an error anyway), and this function should only return early with an error if TPMContext.CreateResourceContextFromTPM returns an error other than tpm2.ResourceUnavailableError.

Reply:
Ok, sorry for that. Feel free to just leave as is if there is no better way to write this.

Reply:
There was quite a bit of duplicated code for creating the 2 primary keys, so I ended up splitting this out in to another helper (provisionPrimaryKey) and I've tried to clean it up a little.